NEP-508: Resharding v2 (#508)
near/NEPs: neps/nep-0508.md

---
NEP: 508
Title: Resharding phase 2
Authors: Waclaw Banasik, Shreyan Gupta, Yoon Hong
Status: Draft
DiscussionsTo: https://github.com/near/nearcore/issues/8992
Type: Protocol
Version: 1.0.0
Created: 2022-09-19
LastUpdated: 2023-09-19
---

> ## Summary

Author: I will add the last point under requirement section as I want to leave summary at the high level without too much technical details

While the new approach addresses critical limitations left unsolved in NEP-40 and is expected to remain valid for foreseable future, it does not serve all usecases, such as dynamic resharding.

## Motivation

Currently, NEAR protocol has four shards. With more partners onboarding, we started seeing that some shards occasionally become over-crowded. In addition, with state sync and stateless validation, validators do not need to track all shards and validator hardware requirements can be greatly reduced with smaller shard size. 

## Specification

### High level assumptions

* Some form of State sync (centralized or decentralized) is enabled.
* Flat state is enabled.
* Shard split boundary is predetermined. In other words, necessity of shard splitting is manually decided.
* Merkle Patricia Trie is undelying data structure for the protocol.
* Minimal epoch gap between two resharding events is X.

### High level requirements

* Resharding should work even when validators stop tracking all shards.
* Resharding should work after stateless validation is enabled.
* Resharding should be fast enough so that both state sync and resharding can happen within one epoch.

> * Resharding should not require additional hardware from nodes.

Reviewer: I believe this is false. There are a few different reasons for increased hardware requirements.
- while resharding we need to store duplicate state for the shard that is being split
- increasing the number of shards naturally increases the total gas per block
- resharding and catchup are resource consuming

Author: I had doubt and am happy that we are having this conversation. If we need additional hardware specs, we need to specify what they should be. I don't think we have a good answer right now, but let's gather the data while testing.

Reviewer: Agreed. I hope the node team can help us benchmark the additional hardware requirements as part of pre-release mocknet testing.

* Resharding should be fault tolerant
  * Chain must not stall in case of resharding failure.

>   * A validator should be able to recover in case they go offline during resharding.

Reviewer: This may mean different things
- the validator can join back immediately and as long as they have enough time they will be fine
- the validator can join back after resharding is finished

* No transaction or receipt should be lost during resharding.
* Resharding should work regardless of number of existing shards.

### Out of scope

* Dynamic resharding

> * Shard determination logic (shard boundary is still determined by string value)

Reviewer: I think dynamic resharding encompasses two things:
- automatically scheduling the resharding
- automatically determining the shard layout 
neither of which we will support in this iteration

> * TBD

Reviewer: merging shard, moving existing boundary (aka only splitting shard is supported)

## Alternatives

* Why is this design the best in the space of possible designs?
  * TBD
* What other designs have been considered and what is the rationale for not choosing them?
  * TBD
* What is the impact of not doing this?
  * TBD

## Future possibilities

As noted above, dynamic resharding is out of scope for this NEP and should be implemented in the future. Dynamic resharding includes the following but not limited to:
* automatic determination of split boundary
* automatic shard splitting and merging based on traffic

Other useful features that can be considered as a follow up:
* account colocation for low latency across account call
* shard on demand

## Consequences

### Positive

* Workload across shards will be more evenly distributed.
* Required space to maintain state (either in memory or in persistent disk) will be smaller.
* State sync overhead will be smaller.
* TBD

### Neutral

* Number of shards is expected to increase.
* Underlying trie structure and data structure are not going to change.

> * Resharding will create dependency on flat storage and state sync.

Reviewer: I wouldn't say that there is a dependency on state sync. Without state sync, while all nodes track all shards, our job would be easier actually.

### Negative

* The resharding process is still not fully automated. Analyzing shard data, determining the split boundary, and triggering an actual shard split all need to be manually curated by a person.

> * During resharding, a node is expected to do more work as it will have to apply changes twice (for the current shard and future shard).

Reviewer: potential to break existing apps and tools that do not handle shard layout changes correctly

Reviewer: Well we'll do our best to notify the community that from now on they need to be able to handle shard layout changing. This NEP is actually part of that communication. It is still a potential breakage so a negative. I don't think we can find every single tool and app on near and personally ask them to prepare.

While the new approach addresses critical limitations left unsolved in NEP-40 and is expected to remain valid for foreseable future, it does not serve all usecases, such as dynamic resharding. 

## Motivation

Currently, NEAR protocol has four shards. With more partners onboarding, we started seeing that some shards occasionally become over-crowded. In addition, with state sync and stateless validation, validators will not need to track all shards and validator hardware requirements can be greatly reduced with smaller shard size. 

## Specification

### High level assumptions

* Flat state is enabled.
* Shard split boundary is predetermined. In other words, necessity of shard splitting is manually decided.
* Merkle Patricia Trie is the undelying data structure for the protocol state.
* Minimal epoch gap between two resharding events is X.

### High level requirements

* Resharding should work even when validators stop tracking all shards.

> * Resharding should work after stateless validation is enabled.

Reviewer 2: Should this be a part of the NEP? We are technically not handling stateless validation.

Author: we can move this to future follow ups?

* Flat storage is enabled.
* Shard split boundary is predetermined and hardcoded. In other words, necessity of shard splitting is manually decided.
* For the time being resharding as an event is only going to happen once but we would still like to have the infrastrcture in place to handle future resharding events with ease.
* Merkle Patricia Trie is the undelying data structure for the protocol state.
* Epoch is at least 6 hrs long for resharding to complete.

### High level requirements

* Resharding must be fast enough so that both state sync and resharding can happen within one epoch.
* Resharding should work efficiently within the limits of the current hardware requirements for nodes.
* Potential failures in resharding may require intervention from node operator to recover.
* No transaction or receipt must be lost during resharding.
* Resharding must work regardless of number of existing shards.
* No apps, tools or code should hardcode the number of shards to 4.

### Out of scope

* Dynamic resharding
  * automatically scheduling resharding based on shard usage/capacity
  * automatically determining the shard layout
* Merging shards or boundary adjustments
* Shard reshuffling

> * Shard Layout determination logic. Shard boundaries are still determined offline and hardcoded.

Reviewer 3: nit: I think it is already mentioned in "Dynamic resharding" section

* In epoch T, in the last block of the epoch, the EpochConfig for epoch T+2 will be set. The EpochConfig for epoch T+2 will have the new shard layout.
* In epoch T + 1, all nodes will perform the state split. The child shards will be kept up to date with the blockchain up until the epoch end first via catchup, and later as part of block postprocessing state application.
* In epoch T + 2, the chain will switch to the new shard layout. 

## Reference Implementation

The implementation heavily re-uses the implementation from [NEP-40](https://github.com/near/NEPs/blob/master/specs/Proposals/0040-split-states.md). Below are listed the major differences and additions. 

### Code pointers to the proposed implementation 

* [new shard layout](https://github.com/near/nearcore/blob/bc0eb3c6607f4ae865526a25b80706ec4e081fdc/core/primitives/src/shard_layout.rs#L161)
* [the main logic for splitting states](https://github.com/near/nearcore/blob/bc0eb3c6607f4ae865526a25b80706ec4e081fdc/chain/chain/src/resharding.rs#L248)
* [the main logic for applying chunks to split states](https://github.com/near/nearcore/blob/bc0eb3c6607f4ae865526a25b80706ec4e081fdc/chain/chain/src/chain.rs#L5180)

> * [the main logic for garbage collecting state from parent shard](https://github.com/near/nearcore/blob/eb824087fdd0763f00e695e303d7ad6c56f96538/chain/chain/src/store.rs#L2325)

Reviewer 3: Links need to be updated, at least for shard layout:  https://github.com/near/nearcore/blob/master/core/primitives/src/shard_layout.rs#L161

Logic for applying chunks was recently moved as well, but it can be moved somewhere else in near future :)

Reviewer 2: It would make sense to use the hash of the latest commit on master instead of the master branch as that can change over time. I think @wacban has already done this in his change.

### Flat Storage

The old implementaion of resharding relied on iterating over the full trie state of the parent shard in order to build the state for the children shards. This implementation was suitable at the time but since then the state has grown considerably and this implementation is now too slow to fit within a single epoch. The new implementation relies on iterating through the flat storage in order to build the children shards quicker. Based on benchmarks, splitting the largest shard by using flat storage can take around 15 min without throttling and around 3 hours with throttling to maintain block production rate.

> The new implementation will also propagate the flat storage for the children shards and keep it up to the with the chain until the switch to the new shard layout in the next epoch. The old implementation didn't handle this case because the flat storage didn't exist back then.

Reviewer 4: nit: grammar near "keep it up to the with the chain"

### New shard layout

The first release of the resharding v2 will contain a new shard layout where one of the existing shards will be split into two smaller shards. Furthermore additional reshardings can be scheduled in subsequent neard releases without additional NEPs unless the need for it arises. A new shard layout can be determined and will be scheduled and executed with the next protocol upgrade. Resharding will typically happen by splitting one of the existing shards into two smaller shards. The new shard layout will be created by adding a new boundary account that will be determined by analysing the storage and gas usage metrics within the shard and selecting a point that will divide the shard roughly in half in accordance to the mentioned metrics. Other metrics can also be used based on requirements.

### Removal of Fixed shards

> Fixed shards was a feature of the protocol that allowed for assigning specific accounts and all of their recursive sub accounts to a predetermined shard. This feature was only used for testing and was never used in production. Fixed shards fature unfortunately breaks the contiguity of shards and is not compatible with the new resharding flow. A sub account of a fixed shard account can fall in the middle of account range that belongs to a different shard. This property of fixed shards made it particularly hard to reason about and implement efficient resharding. 

Reviewer 4: nit: maybe provide a quick example? Just to avoid confusion when someone with no context reads it. Perhaps also quickly mention that having shard colocation gives no benefit in Near.

### Transaction pool

The transaction pool is sharded i.e. it groups transactions by the shard where each transaction should be converted to a receipt. The transaction pool was previously sharded by the ShardId. Unfortunately ShardId is insufficient to correctly identify a shard across a resharding event as ShardIds change domain. The transaction pool was migrated to group transactions by ShardUId instead, and a transaction pool resharding was implemented to reassign transaction from parent shard to children shards right before the new shard layout takes effect. The ShardUId contains the version of the shard layout which allows differentiating between shards in different shard layouts. 

This was implemented ahead of this NEP and the transaction pool is now fully **migrated** to ShardUId. 

## Alternatives

### Why is this design the best in the space of possible designs?

This design is the simplest, most robust and safe while meeting all of the requirements. 

> ### What other designs have been considered and what is the rationale for not choosing them?

Reviewer 4: Perhaps this section should be titled, "What other implementation strategies have been considered ...", since these are not designs. At a high level perhaps the "other designs considered" would include things like, stopping the chain to do an offline resharding and resuming again; or why we decided that we would only split shards and not be able to merge them; why we decided to go with hardcoded resharding boundaries for now.

* When generating proofs for incoming receipts. The proof for each of the children shards contains only the receipts of the shard but it's generated on the parent shard layout and so may not be verified. 

In this NEP we propose that resharding should be rolled out first, before any real dependency on state sync is added. We can then safely roll out the resharding logic and solve the above mentioned issues separately. We believe atleast some of the issues can be mitigated by the implementation of new pre-state root and chunk execution design.

## Integration with Stateless Validation

The Stateless Validation requires that chunk producers provide proof of correctness of the transition function from one state root to another. That proof for the first block after the new shard layout takes place will need to prove that the entire state split was correct as well as the state transition.

> In this NEP we propose that resharding should be rolled out first, before stateless validation. We can then safely roll out the resharding logic and solve the above mentioned issues separately.

Reviewer 4: To hint that we aren't just ignoring the problem, we can mention here that we think a solution is likely feasible by producing a proof of the partial trie before resharding that is enough to validate the resharding, since resharding likely copies over small numbers of large subtries.

## Future fast-followups

### Resharding should work even when validators stop tracking all shards

As mentioned above under 'Integration with State Sync' section, initial release of resharding v2 will happen before the full implementation of state sync and we plan to tackle the integration between resharding and state sync after the next shard split (Won't need a separate NEP as the integration does not require protocol change.)

### Resharding should work after stateless validation is enabled

As mentioned above under 'Integration with Statelss Validation' section, initial release of resharding v2 will happen before the full implementation of stateless validation and we plan to tackle the integration between resharding and stateless validation after the next shard split (May need a separate NEP depending on implemetnation detail.)

> ## Future possibilities

Reviewer 3: Can we add some guidance on how next reshardings can be done? Say, how complex would it be to do resharding from 5 to 6?

### Dynamic resharding

As noted above, dynamic resharding is out of scope for this NEP and should be implemented in the future. Dynamic resharding includes the following but not limited to:

* Automatic determination of split boundary based on parameters like traffic, gas usage, state size, etc.
* Automatic shard splitting and merging

### Localization of resharding event to specific shard

As of today, at the RocksDB storage layer, we have the ShardUId, i.e. the ShardId along with the ShardVersion, as a prefix in the key of trie state and flat state. During a resharding event, we increment the ShardVersion by one, and effectively remap all the current parent shards to new child shards. This implies we can't use the same underlying key value pairs for store and instead would need to duplicate the values with the new ShardUId prefix, even if a shard is unaffected and not split.

In the future, we would like to potentially change the schema in a way such that only the shard that is splitting is impacted by a resharding event, so as to avoid additonal work done by nodes tracking other shards.

### Other useful features

> * Account colocation for low latency across account call

Reviewer 4: I've never heard about this being considered - would it be useful to elaborate a little more?

Also for "Shard on demand" what does that mean?

Reviewer: * colocation - I think this relates to an idea that was dicussed on the side during the Mardid offsite but I'm not 100% sure about the details. 
* shard on demand - I'm not certain what exactly does that mean. 

@walnut-the-cat did you add those? Can you clarify it a bit?

* Removal of shard uids and introducing globally unique shard ids
* Shard on demand

## Consequences

### Positive

* Workload across shards will be more evenly distributed.
* Required space to maintain state (either in memory or in persistent disk) will be smaller. This is useful for in-memory tries.
* State sync overhead will be smaller with smaller state size.

### Neutral

* Number of shards would increase.
* Underlying trie structure and data structure are not going to change.
* Resharding will create dependency on flat state snapshots.

### Negative

> * The resharding process, as of now, is not fully automated. Analyzing shard data, determining the split boundary, and triggering an actual shard split all need to be manually curated and tracked.

Reviewer 4: This is not so much of a negative - as we didn't have this today either.

* During resharding, a node is expected to require more resources as it will first need to copy state data from the parent shard to the child shard, and then will have to apply trie and flat state changes twice, once for the parent shard and once for the child shards.
* Increased potential for apps and tools to break without proper shard layout change handling.

### Backwards Compatibility

> We do not expect anything to break with this change. Yet, shard splitting can introduce additional complexity on replayability. For instance, as target shard of a receipt and belonging shard of an account can change with shard splitting, shard splitting must be replayed along with transactions at the exact epoch boundary.

Reviewer 3: I would add that all tools assuming `num_shards = 4` will break, if tool maintainters monitor NEPs

Reviewer 4: Also, light clients cannot get the shard boundary from the chain itself. The chain itself unfortunately does not encode the shard layout. It only encodes the protocol version, and then we need to figure out the shard layout ourselves. So light clients would have to be updated.

---
NEP: 508
Title: Resharding v2
Authors: Waclaw Banasik, Shreyan Gupta, Yoon Hong
Status: Draft
DiscussionsTo: https://github.com/near/nearcore/issues/8992
Type: Protocol
Version: 1.0.0
Created: 2023-09-19
LastUpdated: 2023-11-14
---

## Summary

This proposal introduces a new implementation for resharding and a new shard layout for the production networks. 

In essence, this NEP is an extension of [NEP-40](https://github.com/near/NEPs/blob/master/specs/Proposals/0040-split-states.md), which was focused on splitting one shard into multiple shards. 

> We are introducing resharding v2, which supports one shard splitting into two within one epoch at a pre-determined split boundary. The NEP includes performance improvement to make resharding feasible under the current state as well as actual resharding in mainnet and testnet (To be specific, spliting the largest shard into two). 

Reviewer 5: ```suggestion
We are introducing resharding v2, which supports one shard splitting into two within one epoch at a pre-determined split boundary. The NEP includes performance improvement to make resharding feasible under the current state as well as actual resharding in mainnet and testnet (To be specific, splitting the largest shard into two). 
```

> While the new approach addresses critical limitations left unsolved in NEP-40 and is expected to remain valid for foreseable future, it does not serve all usecases, such as dynamic resharding. 

Reviewer 5: ```suggestion
While the new approach addresses critical limitations left unsolved in NEP-40 and is expected to remain valid for foreseeable future, it does not serve all use cases, such as dynamic resharding. 
```

## Motivation

Currently, NEAR protocol has four shards. With more partners onboarding, we started seeing that some shards occasionally become over-crowded with respect to total state size and number of transactions. In addition, with state sync and stateless validation, validators will not need to track all shards and validator hardware requirements can be greatly reduced with smaller shard size. With future in-memory tries, it's also important to limit the size of individual shards.

## Specification

### High level assumptions

* Flat storage is enabled.
* Shard split boundary is predetermined and hardcoded. In other words, necessity of shard splitting is manually decided.

> * For the time being resharding as an event is only going to happen once but we would still like to have the infrastrcture in place to handle future resharding events with ease.

Reviewer 5: ```suggestion
* For the time being resharding as an event is only going to happen once but we would still like to have the infrastructure in place to handle future resharding events with ease.
```

> * Merkle Patricia Trie is the undelying data structure for the protocol state.

Reviewer 5: ```suggestion
* Merkle Patricia Trie is the underlying data structure for the protocol state.
```

* Epoch is at least 6 hrs long for resharding to complete.

### High level requirements

* Resharding must be fast enough so that both state sync and resharding can happen within one epoch.
* Resharding should work efficiently within the limits of the current hardware requirements for nodes.
* Potential failures in resharding may require intervention from node operator to recover.
* No transaction or receipt must be lost during resharding.
* Resharding must work regardless of number of existing shards.
* No apps, tools or code should hardcode the number of shards to 4.

### Out of scope

* Dynamic resharding
  * automatically scheduling resharding based on shard usage/capacity
  * automatically determining the shard layout
* Merging shards or boundary adjustments
* Shard reshuffling

### Required protocol changes

A new protocol version will be introduced specifying the new shard layout which would be picked up by the resharding logic to split the shard.

### Required state changes

> * For the duration of the resharding the node will need to maintain a snapshot of the flat state and related columns. As the main database and the snapshot diverge this will cause some extent of storage overhead.

Reviewer 6: is there an estimate of how much more storage is required during resharding

Reviewer: Added rounded up numbers based on [results shared on zulip](https://near.zulipchat.com/#narrow/stream/407288-pagoda.2Fcore.2Fresharding/topic/hardware.20overhead)

* For the duration of the epoch before the new shard layout takes effect, the node will need to maintain the state and flat state of shards in the old and new layout at the same time. The State and FlatState columns will grow up to approx 2x the size. The processing overhead should be minimal as the chunks will still be executed only on the parent shards. There will be increased load on the database while applying changes to both the parent and the children shards. 

### Resharding flow

> * The new shard layout will be agreed on offline by the protocol team and hardcoded in the reference implementation.

Reviewer 6: Given that this is finalized, could we also mention the shard layout here?

Reviewer: I believe @walnut-the-cat wanted to keep the NEP more generic and not mention the next shard layout. I don't have any strong opinions here, I'm happy to do either. @walnut-the-cat are you happy for me to add it? What was the reason for not having it here?

> * The new shard layout will be agreed on offline by the protocol team and hardcoded in the neard reference implementation.

Reviewer 5: ```suggestion
* The new shard layout will be agreed on offline by the protocol team and hardcoded in the reference implementation.
```

* In epoch T + 2, the chain will switch to the new shard layout. 

## Reference Implementation

The implementation heavily re-uses the implementation from [NEP-40](https://github.com/near/NEPs/blob/master/specs/Proposals/0040-split-states.md). Below are listed the major differences and additions. 

### Code pointers to the proposed implementation 

* [new shard layout](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/core/primitives/src/shard_layout.rs#L161)  
* [the main logic for splitting states](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/chain/chain/src/resharding.rs#L280)
* [the main logic for applying chunks to split states](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/chain/chain/src/update_shard.rs#L315)
* [the main logic for garbage collecting state from parent shard](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/chain/chain/src/store.rs#L2335)

### Flat Storage

> The old implementaion of resharding relied on iterating over the full trie state of the parent shard in order to build the state for the children shards. This implementation was suitable at the time but since then the state has grown considerably and this implementation is now too slow to fit within a single epoch. The new implementation relies on iterating through the flat storage in order to build the children shards quicker. Based on benchmarks, splitting the largest shard by using flat storage can take around 15 min without throttling and around 3 hours with throttling to maintain block production rate.

Reviewer 5: ```suggestion
The old implementation of resharding relied on iterating over the full trie state of the parent shard in order to build the state for the children shards. This implementation was suitable at the time but since then the state has grown considerably and this implementation is now too slow to fit within a single epoch. The new implementation relies on iterating through the flat storage in order to build the children shards quicker. Based on benchmarks, splitting the largest shard by using flat storage can take around 15 min without throttling and around 3 hours with throttling to maintain the block production rate.
```

> The old implementation of resharding relied on iterating over the full trie state of the parent shard in order to build the state for the children shards. This implementation was suitable at the time but since then the state has grown considerably and this implementation is now too slow to fit within a single epoch. The new implementation relies on iterating through the flat storage in order to build the children shards quicker. Based on benchmarks, splitting the largest shard by using flat storage can take around 15 min without throttling and around 3 hours with throttling to maintain the block production rate.

Reviewer 6: Unrelated to this proposal, but do I understand correctly that once we have in-memory trie, the old approach would actually work again?

Reviewer: Hmm good question. One requirement that we have for storage is to give us access to a frozen view of the chain at a pre-defined block. This is why we need the flat store snapshots. I'm not sure if such snapshottting would be feasible with mem-trie. It's definitely worth exploring as a number of fundamental assumptions will change.

Reviewer 2: Although theoretically it does seem plausible given we have a large enough GC period for mem-trie and we maintain "historic" state for the GC period.

Reviewer 6: > One requirement that we have for storage is to give us access to a frozen view of the chain at a pre-defined block. This is why we need the flat store snapshots. I'm not sure if such snapshottting would be feasible with mem-trie

My understanding is that in-memory trie just moves trie to be held in memory and does not change its persistence assumptions, so yes you can read the state at some predefined block (assuming it has not been garbage collected).

Reviewer 4: In-memory trie loads from flat storage and indeed does not change any persistence assumptions. If there is flat storage available at some block (such as via a flat state snapshot) then memtries can be loaded that corresponds to that flat state. It's possible to then iterate through the memtrie to reconstruct a resharded on-disk trie or flat state. That being said, this does not sound so much like "the old approach works again", but rather that memtries can be used as an optimization to the (pretty isolated) step of reconstructing a trie from flat state -- which is the solution we're adopting in this proposal. We would be doing the same thing, except that memtrie construction from flat state is a bit more efficient than direct trie reconstruction of flat state (for reasons of doing less wasted work and doing less disk I/O).

The pros would be that this is more efficient (perhaps 2x-4x more would be my ballpark guess), the cons is that this requires more memory and, well, we don't have the code for dumping memtrie into on-disk trie. It sounds easy, but they are different data structures and in practice we have to carefully deal with refcount keeping which will not be trivial.

The new implementation will also propagate the flat storage for the children shards and keep it up to date with the chain until the switch to the new shard layout in the next epoch. The old implementation didn't handle this case because the flat storage didn't exist back then.

In order to ensure consistent view of the flat storage while splitting the state the node will maintain a snapshot of the flat state and related columns as of the last block of the epoch prior to resharding. The existing implementation of flat state snapshots used in State Sync will be used for this purpose.

### Handling receipts, gas burnt and balance burnt

> When resharding, extra care should be taken when handling receipts in order to ensure that no receipts are lost or duplicated. The gas burnt and balance burnt also need to be correclty handled. The old resharding implementation for handling receipts, gas burnt and balance burnt relied on the fact in the first resharding there was only a single parent shard to begin with. The new implementation will provide a more generic and robust way of reassigning the receipts to the child shards, gas burnt, and balance burnt, that works for arbitrary splitting of shards, regardless of the previous shard layout. 

Reviewer 5: ```suggestion
When resharding, extra care should be taken when handling receipts in order to ensure that no receipts are lost or duplicated. The gas burnt and balance burnt also need to be correctly handled. The old resharding implementation for handling receipts, gas burnt and balance burnt relied on the fact in the first resharding there was only a single parent shard to begin with. The new implementation will provide a more generic and robust way of reassigning the receipts to the child shards, gas burnt, and balance burnt, that works for arbitrary splitting of shards, regardless of the previous shard layout. 
```

### New shard layout

> The first release of the resharding v2 will contain a new shard layout where one of the existing shards will be split into two smaller shards. Furthermore additional reshardings can be scheduled in subsequent neard releases without additional NEPs unless the need for it arises. A new shard layout can be determined and will be scheduled and executed with the next protocol upgrade. Resharding will typically happen by splitting one of the existing shards into two smaller shards. The new shard layout will be created by adding a new boundary account that will be determined by analysing the storage and gas usage metrics within the shard and selecting a point that will divide the shard roughly in half in accordance to the mentioned metrics. Other metrics can also be used based on requirements.

Reviewer 5: ```suggestion
The first release of the resharding v2 will contain a new shard layout where one of the existing shards will be split into two smaller shards. Furthermore additional reshardings can be scheduled in subsequent releases without additional NEPs unless the need for it arises. A new shard layout can be determined and will be scheduled and executed with the next protocol upgrade. Resharding will typically happen by splitting one of the existing shards into two smaller shards. The new shard layout will be created by adding a new boundary account that will be determined by analysing the storage and gas usage metrics within the shard and selecting a point that will divide the shard roughly in half in accordance to the mentioned metrics. Other metrics can also be used based on requirements.
```

### Removal of Fixed shards

Fixed shards was a feature of the protocol that allowed for assigning specific accounts and all of their recursive sub accounts to a predetermined shard. This feature was only used for testing and was never used in production. Fixed shards feature unfortunately breaks the contiguity of shards and is not compatible with the new resharding flow. A sub account of a fixed shard account can fall in the middle of account range that belongs to a different shard. This property of fixed shards made it particularly hard to reason about and implement efficient resharding. 

For example in a shard layout with boundary accounts [`b`, `d`] the account space is cleanly divided into three shards, each spanning a contiguous range and account ids:

* 0 -  `:b`
* 1 - `b:d`
* 2 - `d:`

Now if we add a fixed shard `f` to the same shard layout, then any we'll have 4 shards but neither is contiguous. Accounts such as `aaa.f`, `ccc.f`, `eee.f` that would otherwise belong to shards 0, 1 and 2 respectively are now all assigned to the fixed shard and create holes in the shard account ranges.  

> It's also worth noting that there is no benefit to having accounts colocated in the same shard. Any transaction or receipts is treated the same way regardless of crossing shard boundary. 

Reviewer 5: ```suggestion
It's also worth noting that there is no benefit to having accounts colocated in the same shard. Any transaction or receipt is treated the same way regardless of crossing shard boundary. 
```

This was implemented ahead of this NEP and the fixed shards feature was **removed**. 

### Garbage collection

> In epoch T+2 once resharding is completed, we can delete the trie state and the flat state related to the parent shard. In practice, this is handled as part of the garbage collection code. While garbage collecting the last block of epoch T+1, we go ahead and clear all the data associated with the parent shard from the trie cache, flat storage, and RocksDB state associated with trie state and flat storage.

Reviewer 6: Does this mean that the storage increase would last for 5 epochs (or however long the garbage collection period is configured to be)

Reviewer 2: Yes, that's correct, we maintain the state of the parent shard for the period of GC. Note that right after resharding, we keep deleting state from the `State` (Trie data) column of the parent shard with GC normally while new state is being added to the `State` column of the child shards upto the GC period. This means the storage increase compared to a non-resharding case would amortized be the size of the current state as of the time of resharding and not all the historic data.

### Transaction pool

The transaction pool is sharded i.e. it groups transactions by the shard where each transaction should be converted to a receipt. The transaction pool was previously sharded by the ShardId. Unfortunately ShardId is insufficient to correctly identify a shard across a resharding event as ShardIds change domain. The transaction pool was migrated to group transactions by ShardUId instead, and a transaction pool resharding was implemented to reassign transaction from parent shard to children shards right before the new shard layout takes effect. The ShardUId contains the version of the shard layout which allows differentiating between shards in different shard layouts. 

This was implemented ahead of this NEP and the transaction pool is now fully **migrated** to ShardUId. 

## Alternatives

### Why is this design the best in the space of possible designs?

> This design is the simplest, most robust and safe while meeting all of the requirements. 

Reviewer 5: ```suggestion
This design is simple, robust, safe, and meets all requirements. 
```

### What other designs have been considered and what is the rationale for not choosing them?

#### Alternative implementations

* Splitting the trie by iterating over the boundaries between children shards for each trie record type. This implementation has the potential to be faster but it is more complex and it would take longer to implement. We opted in for the much simpler one using flat storage given it is already quite performant.
* Changing the trie structure to have the account id first and type of record later. This change would allow for much faster resharding by only iterating over the nodes on the boundary. This approach has two major drawbacks without providing too many benefits over the previous approach of splitting by each trie record type.
  1) It would require a massive migration of trie.

>   2) We would need to maintain the old and the new trie structure forever.

Reviewer 5: > Changing the trie structure to have the account id first and type of record later.

What is the `type of record`?

> We would need to maintain the old and the new trie structure forever.

Out of curiosity, why would we need to maintain the old structure?

Reviewer: > What is the `type of record`?

The trie is an implementation of a key-value storage that supports multiple different value types. A single account may own multiple key-values, or records in the trie. For example an account with a deployed contract may have one record for storing the Account data, one record for storing the ContractCode and multiple records for storing ContractData. 

Here is the list of all possible types:
https://github.com/near/nearcore/blob/abf6e99bad479a216a7f6e8a465a3eaf0523659e/core/primitives/src/trie_key.rs#L16

> Out of curiosity, why would we need to maintain the old structure? 

The old structure was commited into the blockchain in the form of the state root merkle hash of the trie. Changing the trie structure would change the merkle hash. In order to be able to deal with historical data we would need to keep the old implementation in the code. 

Arguably this may only be needed for archival nodes and RPC nodes would need to maintain both only for a transitional period. In practice we would likely be forced to keep it foreve unless willing to fork the neard implementation for archival nodes.

Reviewer: By committed to the blockchain I mean that the state roots were included in the headers that were then signed by producers and approved by validators.

  * The extended adjustments are currently not justified. Both merging and boundary moving may be useful in the future when the traffic patterns change and some shard become underutilized. In the nearest future we only predict needing to reduce the size of the heaviest shards. 

### What is the impact of not doing this?

We need resharding in order to scale up the system. Without resharding eventually shards would grow so big (in either storage or cpu usage) that a single node would not be able to handle it. Additionally, this clears up the path to implement in-memory tries as we need to store the whole trie structure in limited RAM. In the future smaller shard size would lead to faster syncing of shard data when nodes start tracking just one shard.

## Integration with State Sync

There are two known issues in the integration of resharding and state sync:

* When syncing the state for the first epoch where the new shard layout is used. In this case the node would need to apply the last block of the previous epoch. It cannot be done on the children shard as on chain the block was applied on the parent shards and the trie related gas costs would be different.

> * When generating proofs for incoming receipts. The proof for each of the children shards contains only the receipts of the shard but it's generated on the parent shard layout and so may not be verified. 

Reviewer 7: @karim-en will this have an impact on the bridge?

Reviewer 8: It looks like yes, but I don't understand how this could impact the Prover logic, so I would like to see the structure of the newly generated proof.
Also, it would be good if @walnut-the-cat had a look into our proof verification code and gave us some feedback.
https://github.com/aurora-is-near/rainbow-bridge/blob/bf7cc1e81ee4a4fad3e74e2c599b0b6cf3757f4d/contracts/eth/nearprover/contracts/NearProver.sol#L28

Reviewer: @birchmd and @karim-en I had a quick look but I completely lack context. Perhaps it would be easier to perform an end to end test of the bridge and resharding on localnet or other small network? A simple setup to trigger resharding would be as follows:
* prepare custom binary with resharding stabilized e.g. like so https://github.com/near/nearcore/commit/fdcfedf01d296358aa1146ade6b7e723a31ae0cb 
* run 1.36 for an epoch or two
* stop the node(s)
* run the custom binary for at least three epochs
* assuming the binary upgrade happens in epoch T, the resharding will happen in T+1, and the network will switch to the new shard layout in the first block of T+2

Please let me know if you need any help at all and we can look into it together.

> In this NEP we propose that resharding should be rolled out first, before any real dependency on state sync is added. We can then safely roll out the resharding logic and solve the above mentioned issues separately. We believe atleast some of the issues can be mitigated by the implementation of new pre-state root and chunk execution design.

Reviewer 5: ```suggestion
In this NEP we propose that resharding should be rolled out first, before any real dependency on state sync is added. We can then safely roll out the resharding logic and solve the above mentioned issues separately. We believe at least some of the issues can be mitigated by the implementation of new pre-state root and chunk execution design.
```

In this NEP we propose that resharding should be rolled out first, before stateless validation. We can then safely roll out the resharding logic and solve the above mentioned issues separately. This issue was discussed with the stateless validation experts and we are cautiously optimistic that the integration will be possible. The most concerning part is the proof size and we believe that it should be small enough thanks to the resharding touching relatively small number of trie nodes - on the order of the depth of the trie. 

## Future fast-followups

### Resharding should work even when validators stop tracking all shards

As mentioned above under 'Integration with State Sync' section, initial release of resharding v2 will happen before the full implementation of state sync and we plan to tackle the integration between resharding and state sync after the next shard split (Won't need a separate NEP as the integration does not require protocol change.)

### Resharding should work after stateless validation is enabled

> As mentioned above under 'Integration with Statelss Validation' section, initial release of resharding v2 will happen before the full implementation of stateless validation and we plan to tackle the integration between resharding and stateless validation after the next shard split (May need a separate NEP depending on implemetnation detail.)

Reviewer 5: ```suggestion
As mentioned above under 'Integration with Stateless Validation' section, the initial release of resharding v2 will happen before the full implementation of stateless validation and we plan to tackle the integration between resharding and stateless validation after the next shard split (May need a separate NEP depending on implementation detail.)
```

## Consequences

### Positive

* Workload across shards will be more evenly distributed.
* Required space to maintain state (either in memory or in persistent disk) will be smaller. This is useful for in-memory tries.
* State sync overhead will be smaller with smaller state size.

### Neutral

* Number of shards would increase.
* Underlying trie structure and data structure are not going to change.
* Resharding will create dependency on flat state snapshots.
* The resharding process, as of now, is not fully automated. Analyzing shard data, determining the split boundary, and triggering an actual shard split all need to be manually curated and tracked.

### Negative

* During resharding, a node is expected to require more resources as it will first need to copy state data from the parent shard to the child shard, and then will have to apply trie and flat state changes twice, once for the parent shard and once for the child shards.
* Increased potential for apps and tools to break without proper shard layout change handling.

### Backwards Compatibility

> Any light clients, tooling or frameworks external to nearcore that have the current shard layout or the current number of shards hardcoded may break and will need to be adjusted in advance. The recommended way for fixing it is querying an RPC node for the shard layout of the relevant epoch and using that information in place of the previously hardcoded shard layout or number of shards. 

Reviewer 7: It might be nice to show an example of this query.

As far as I know, the main reason to have a hard-coded shard id is to query transactions from a specific account by downloading just the chunk that pertains to that account. This can be done using the `block_id` + `shard_id` version of the [chunk details query](https://docs.near.org/api/rpc/block-chunk#chunk-details).

After a split, where an account might move to a larger shard id, how should a tool recover? How will the tool know that it needs to change which shard id it is querying? If the tool is simply iterating over the chunk looking for a specific account, it will not be able to distinguish a chunk with happens to not including any transactions from that account from the situation where a split just happened and there are transactions now located in a different chunk?

Reviewer: Thanks! Can you share an example of an implementation of this pattern? The adjusted pattern may look something like the following. I'll implement the new RPC endpoints soon and update the NEP to reflect that. 
- shard_layout := rpc(method="shard_layout", block_hash=block_hash)
- shard_id = shard_layout.get_shard_id_for_account(account)
- ... the rest same as before

This would be the least intrusive change but frankly quering the whole chunk just to get transactions for a single account could be done with a dedicated rpc endpoint. This is outside of the scope of this NEP though and we'll focus on delivering the relavant endpoints.

Reviewer 7: I don't know if there is a production implementation of this pattern somewhere, but this kind of query is something I have done myself in the past when looking at recent transactions coming from `relay.aurora` for debugging purposes. The pseudocode looked something like

```
AURORA_RELAYER_SHARD_ID = 3

block_height = read_from_stdin()
chunk = rpc(method="chunk", block_id=block_height, shard_id=AURORA_RELAYER_SHARD_ID)
txs = [t for t in chunk.transactions if t.signer_id == "relay.aurora"]
print(txs)
```

Your proposed RPC method where the shard layout can be directly queried would allow getting rid of the hard-coded shard id and I agree it is the least intrusive.

Reviewer: Added instructions on how to get the current shard layout by using the rpc endpoint EXPERIMENTAL_protocol_config.

TBD. e.g. configuration changes we have to introduce

A new protocol version will be introduced specifying the new shard layout. 

### Required state changes

TBD. e.g. additional/updated data a node has to maintain

* For the duration of the resharding the node will need to maintain a snapshot of the flat state and related columns.

> * For the duration of the epoch before the new shard layout takes effect, the node will need to maintain the state and flat state of shards in the old and new layout at the same time. 

Reviewer: I don't have a concrete number but let me write down what is expected to happen.

### Resharding flow

TBD. how resharding happens at the high level

* The new shard layout will be agreed on offline by the protocol team and hardcoded in the neard reference implementation.
* In epoch T the protocol version upgrade date will pass and nodes will vote to switch to the new protocol version. The new protocol version will contain the new shard layout. 

> * In epoch T, in the last block of the epoch, the EpochConfig for epoch T+2 will be set. The EpochConfig for epoch T+2 will have the new shard layout.

Reviewer: yes

* In epoch T + 1, all nodes will perform the state split. The child shards will be kept up to date with the blockchain up until the epoch end.
* In epoch T + 2, the chain will switch to the new shard layout. 

> ## Reference Implementation

Author: may want to add some code snippet (OR link to the code) in case reviewers want to know more about actual implementation details.

The new implementation will also propagate the flat storage for the children shards and keep it up to the with the chain until the switch to the new shard layout. The old implementation didn't handle this case because the flat storage didn't exist back then. 

In order to ensure consistent view of the flat storage while splitting the state the node will maintain a snapshot of the flat state and related columns. The existing implementation of flat state snapshots used in State Sync will be adjusted for this purpose. 

### Handling receipts, gas burnt and balance burnt

When resharding, extra care should be taken when handling receipts in order to ensure that no receipts are lost or duplicated. The gas burnt and balance burnt also need to be correclty handled. The old resharding implementation for handling receipts, gas burnt and balance burnt relied on the fact in the first resharding there was only a single parent shard to begin with. The new implementation will provide a more generic and robust way of reassigning the receipts, gas burnt and balance burnt that works for arbitrary splitting of shards, regardless of the previous shard layout. 

### New shard layout

> A new shard layout will be determined and will be scheduled and executed in the production networks. The new shard layout will maintain the same boundaries for shards 0, 1 and 2. The heaviest shard today - Shard 3 will be split by introducing a new boundary account. The new boundary account will be determined by analysis the storage and gas usage within the shard and selecting a point that will divide the shard roughly in half in accordance to the mentioned metrics. Other metrics can also be used. 

Author: Let's avoid mentioning a specific target shard as resharding v2 can be used for further shard split, but focus on how new shard boundary/layout will be determined

### Fixed shards

> Fixed shards is a feature of the protocol that allows for assigning specific accounts and all of their recursive sub accounts to a predetermined shard. This feature is only used for testing, it was never used in production and there is no need for it in production. This feature unfortunately breaks the contiguity of shards. A sub account of a fixed shard account can fall in the middle of account range that belongs to a different shard. This property of fixed shards makes it particularly hard to reason about and implement efficient resharding. In order to simplify the code and new resharding implementation the fixed shards feature was removed ahead of this NEP. 

Author: let's bold **removed**

### Transaction pool

> The transaction pool is sharded e.i. it groups transactions by the shard where each should be converted to a receipt. The transaction pool was previously sharded by the ShardId. Unfortunately ShardId is insufficient to correctly identify a shard across a resharding event as ShardIds change domain. The transaction pool was migrated to group transactions by ShardUId instead and a transaction pool resharding was implemented to reassign transaction from parent shard to children shards right before the new shard layout takes effect. This was implemented ahead of this NEP. 

Author: nit: should we mention how shardUID is different from shardID and how using UID instead solves the problem?
